fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentType.java

```diff
@@ -0,0 +1,211 @@
+
+/**
+ * Data attachments allow anyone to "attach" arbitrary data to common game objects (block entities, chunks, entities and world at the moment).
+ *
+ * <p>To attach data, one needs to create an attachment type that specifies the type of the arbitrary data, and how it should be persisted.
+ * This is done using the various {@code for*} factory methods:
+ * <ul>
+ *     <li>{@link #forBlockEntity} for block entities.</li>
+ *     <li>{@link #forChunk} for chunks.</li>
+ *     <li>{@link #forEntity} for entities.</li>
+ *     <li>{@link #forWorld} for worlds.</li>
+ * </ul>
+ *
+ * @param <A> Type of the attached object.
+ * @param <T> Type of the host of this attachment.
+ */
+@ApiStatus.NonExtendable
+public interface AttachmentType<A, T> {
+	/**
+	 * Create a new block entity attachment type with a given identifier, an attachment class, and optionally a serializer.
+	 *
+	 * <p>Serialization will only be used on the logical server.
+	 * Make sure to call {@link BlockEntity#markDirty()} after a change to the data attachments to ensure that they get persisted.
+	 *
+	 * @throws IllegalArgumentException If an attachment type with the given identifier already exists for block entities.
+	 */
+	static <A> AttachmentType<A, BlockEntity> forBlockEntity(Identifier identifier, Class<A> attachmentClass, @Nullable Serializer<A, ? super BlockEntity> serializer) {
+		return AttachmentTypeImpl.create(identifier, attachmentClass, BlockEntity.class, serializer);
+	}
+
+	/**
+	 * Create a new block entity attachment type with a given identifier, an attachment class, and optionally a serializer.
+	 *
+	 * <p>Serialization will only be used on the logical server.
+	 * Make sure to call {@link WorldChunk#setNeedsSaving} after a change to the data attachments to ensure that they get persisted.
+	 *
+	 * @throws IllegalArgumentException If an attachment type with the given identifier already exists for block entities.
+	 */
+	static <A> AttachmentType<A, WorldChunk> forChunk(Identifier identifier, Class<A> attachmentClass, @Nullable Serializer<A, ? super WorldChunk> serializer) {
+		return AttachmentTypeImpl.create(identifier, attachmentClass, WorldChunk.class, serializer);
+	}
+
+	/**
+	 * Create a new block entity attachment type with a given identifier, an attachment class, and optionally a serializer.
+	 *
+	 * <p>Serialization will only be used on the logical server.
+	 * Entity data is always persisted, there is no need to call any method after modifying the attachments.
+	 *
+	 * <p>Note that data is not transferred automatically in the following cases, and requires special handling if desired:
+	 * <ul>
+	 *     <li>On the (logical) client and server: when a player dies and respawns.
+	 *     {@code ServerPlayerEvents.COPY_FROM} can be used to copy the data on the server.</li>
+	 *     <li>On the client: when a player changes dimensions.</li>
+	 *     <li>On the client and server: when a non-player entity changes dimensions.</li>
+	 * </ul>
```

Comment:
Not a fan of this, maybe we should add ways to make this copying easier? I fear that it might make the implementation a lot heavier though.

Reply:
I think without this, its going to cause a number of unexpected bugs. Its not trivial for mods to handle all of these cases on their own either.

I would copy by default, and let the mods remove the data if they dont want that.

Reply:
We need some sort of copy handler, because the attachments might capture a reference to the Entity. I can add that as an additional parameter to `forEntity`.

Reply:
It turns out that serializable attachments are handled correctly. Player deaths require custom handling using the `COPY_FROM` event, which is imo good (you may not want to keep e.g. mana if the player dies).